File: repos/openjarvis/src/openjarvis/security/guardrails.py
"""GuardrailsEngine — security-aware inference engine wrapper."""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import Any, Dict, List, Optional, Sequence

from openjarvis.core.events import EventBus, EventType
from openjarvis.core.types import Message
from openjarvis.engine._stubs import InferenceEngine
from openjarvis.security._stubs import BaseScanner
from openjarvis.security.scanner import PIIScanner, SecretScanner
from openjarvis.security.types import RedactionMode, ScanResult


class SecurityBlockError(Exception):
    """Raised when mode is BLOCK and security findings are detected."""


class GuardrailsEngine(InferenceEngine):
    """Wraps an existing ``InferenceEngine`` with security scanning.

    Not registered in ``EngineRegistry`` — instantiated dynamically to wrap
    any engine at runtime.

    Parameters
    ----------
    engine:
        The wrapped inference engine.
    scanners:
        List of scanners to run.  Defaults to ``SecretScanner`` + ``PIIScanner``.
    mode:
        Action taken on findings: WARN, REDACT, or BLOCK.
    scan_input:
        Whether to scan input messages.
    scan_output:
        Whether to scan output content.
    bus:
        Optional event bus for publishing security events.
    """

    def __init__(
        self,
        engine: InferenceEngine,
        *,
        scanners: Optional[List[BaseScanner]] = None,
        mode: RedactionMode = RedactionMode.WARN,
        scan_input: bool = True,
        scan_output: bool = True,
        bus: Optional[EventBus] = None,
    ) -> None:
        self._engine = engine
        self._scanners: List[BaseScanner] = scanners if scanners is not None else [
            SecretScanner(),
            PIIScanner(),
        ]
        self._mode = mode
        self._scan_input = scan_input
        self._scan_output = scan_output
        self._bus = bus

    # -- properties ----------------------------------------------------------

    @property
    def engine_id(self) -> str:  # type: ignore[override]
        """Delegate to the wrapped engine."""
        return self._engine.engine_id

    # -- scanning helpers ----------------------------------------------------

    def _scan_text(self, text: str) -> ScanResult:
        """Run all scanners on *text* and merge findings."""
        merged = ScanResult()
        for scanner in self._scanners:
            result = scanner.scan(text)
            merged.findings.extend(result.findings)
        return merged

    def _redact_text(self, text: str) -> str:
        """Run all scanners' redact() on *text*."""
        result = text
        for scanner in self._scanners:
            result = scanner.redact(result)
        return result

    def _handle_findings(
        self,
        text: str,
        result: ScanResult,
        direction: str,
    ) -> str:
        """Apply the configured mode to findings.

        Parameters
        ----------
        text:
            The original text.
        result:
            Scan result containing findings.
        direction:
            ``"input"`` or ``"output"`` — used in event data.

        Returns
        -------
        str
            Possibly modified text (unchanged for WARN, redacted for REDACT).

        Raises
        ------
        SecurityBlockError
            If mode is BLOCK.
        """
        finding_dicts = [
            {
                "pattern": f.pattern_name,
                "threat": f.threat_level.value,
                "description": f.description,
            }
            for f in result.findings
        ]

        if self._mode == RedactionMode.WARN:
            if self._bus:
                self._bus.publish(
                    EventType.SECURITY_ALERT,
                    {
                        "direction": direction,
                        "findings": finding_dicts,
                        "mode": "warn",
                    },
                )
            return text

        if self._mode == RedactionMode.REDACT:
            if self._bus:
                self._bus.publish(
                    EventType.SECURITY_ALERT,
                    {
                        "direction": direction,
                        "findings": finding_dicts,
                        "mode": "redact",
                    },
                )
            return self._redact_text(text)

        # BLOCK mode
        if self._bus:
            self._bus.publish(
                EventType.SECURITY_BLOCK,
                {
                    "direction": direction,
                    "findings": finding_dicts,
                    "mode": "block",
                },
            )
        raise SecurityBlockError(
            f"Security scan blocked {direction}: "
            f"{len(result.findings)} finding(s) detected"
        )

    # -- InferenceEngine interface -------------------------------------------

    def generate(
        self,
        messages: Sequence[Message],
        *,
        model: str,
        temperature: float = 0.7,
        max_tokens: int = 1024,
        **kwargs: Any,
    ) -> Dict[str, Any]:
        """Scan input, call wrapped engine, scan output."""
        # Scan input messages
        if self._scan_input:
            processed = list(messages)
            for i, msg in enumerate(processed):
                if msg.content:
                    result = self._scan_text(msg.content)
                    if not result.clean:
                        processed[i] = Message(
                            role=msg.role,
                            content=self._handle_findings(
                                msg.content, result, "input",
                            ),
                            name=msg.name,
                            tool_calls=msg.tool_calls,
                            tool_call_id=msg.tool_call_id,
                            metadata=msg.metadata,
                        )
            messages = processed

        # Call wrapped engine
        response = self._engine.generate(
            messages, model=model, temperature=temperature,
            max_tokens=max_tokens, **kwargs,
        )

        # Scan output
        if self._scan_output:
            content = response.get("content", "")
            if content:
                result = self._scan_text(content)
                if not result.clean:
                    response["content"] = self._handle_findings(
                        content, result, "output"
                    )

        return response

    async def stream(
        self,
        messages: Sequence[Message],
        *,
        model: str,
        temperature: float = 0.7,
        max_tokens: int = 1024,
        **kwargs: Any,
    ) -> AsyncIterator[str]:
        """Yield tokens in real-time, scan accumulated output post-hoc."""
        accumulated = []
        async for token in self._engine.stream(
            messages, model=model, temperature=temperature,
            max_tokens=max_tokens, **kwargs,
        ):
            accumulated.append(token)
            yield token

        # Post-hoc scan of accumulated output for logging only
        if self._scan_output:
            full_output = "".join(accumulated)
            if full_output:
                result = self._scan_text(full_output)
                if not result.clean and self._bus:
                    finding_dicts = [
                        {
                            "pattern": f.pattern_name,
                            "threat": f.threat_level.value,
                            "description": f.description,
                        }
                        for f in result.findings
                    ]
                    self._bus.publish(
                        EventType.SECURITY_ALERT,
                        {
                            "direction": "output",
                            "findings": finding_dicts,
                            "mode": "stream_post_hoc",
                        },
                    )

    def list_models(self) -> List[str]:
        """Delegate to wrapped engine."""
        return self._engine.list_models()

    def health(self) -> bool:
        """Delegate to wrapped engine."""
        return self._engine.health()


__all__ = ["GuardrailsEngine", "SecurityBlockError"]

File: repos/openjarvis/rust/crates/openjarvis-security/src/guardrails.rs
//! GuardrailsEngine — security-aware inference engine wrapper.

use crate::scanner::{PIIScanner, SecretScanner};
use crate::types::{RedactionMode, ScanResult};
use openjarvis_core::error::OpenJarvisError;
use openjarvis_core::{EventBus, EventType, GenerateResult, Message};
use openjarvis_engine::traits::{InferenceEngine, TokenStream};
use serde_json::Value;
use std::sync::Arc;

/// Wraps an existing `InferenceEngine` with security scanning on I/O.
///
/// Generic over `E` for static dispatch when the engine type is known.
pub struct GuardrailsEngine<E: InferenceEngine> {
    engine: E,
    secret_scanner: SecretScanner,
    pii_scanner: PIIScanner,
    mode: RedactionMode,
    scan_input: bool,
    scan_output: bool,
    bus: Option<Arc<EventBus>>,
}

impl<E: InferenceEngine> GuardrailsEngine<E> {
    pub fn new(
        engine: E,
        mode: RedactionMode,
        scan_input: bool,
        scan_output: bool,
        bus: Option<Arc<EventBus>>,
    ) -> Self {
        Self {
            engine,
            secret_scanner: SecretScanner::new(),
            pii_scanner: PIIScanner::new(),
            mode,
            scan_input,
            scan_output,
            bus,
        }
    }

    fn scan_text(&self, text: &str) -> ScanResult {
        let mut result = self.secret_scanner.scan(text);
        let pii_result = self.pii_scanner.scan(text);
        result.findings.extend(pii_result.findings);
        result
    }

    fn redact_text(&self, text: &str) -> String {
        let r = self.secret_scanner.redact(text);
        self.pii_scanner.redact(&r)
    }

    fn handle_findings(
        &self,
        text: &str,
        result: &ScanResult,
        direction: &str,
    ) -> Result<String, OpenJarvisError> {
        let finding_dicts: Vec<Value> = result
            .findings
            .iter()
            .map(|f| {
                serde_json::json!({
                    "pattern": f.pattern_name,
                    "threat": f.threat_level.to_string(),
                    "description": f.description,
                })
            })
            .collect();

        match self.mode {
            RedactionMode::Warn => {
                if let Some(ref bus) = self.bus {
                    let mut data = std::collections::HashMap::new();
                    data.insert(
                        "direction".to_string(),
                        Value::String(direction.to_string()),
                    );
                    data.insert("findings".to_string(), Value::Array(finding_dicts));
                    data.insert("mode".to_string(), Value::String("warn".to_string()));
                    bus.publish(EventType::SecurityAlert, data);
                }
                Ok(text.to_string())
            }
            RedactionMode::Redact => {
                if let Some(ref bus) = self.bus {
                    let mut data = std::collections::HashMap::new();
                    data.insert(
                        "direction".to_string(),
                        Value::String(direction.to_string()),
                    );
                    data.insert("findings".to_string(), Value::Array(finding_dicts));
                    data.insert(
                        "mode".to_string(),
                        Value::String("redact".to_string()),
                    );
                    bus.publish(EventType::SecurityAlert, data);
                }
                Ok(self.redact_text(text))
            }
            RedactionMode::Block => {
                if let Some(ref bus) = self.bus {
                    let mut data = std::collections::HashMap::new();
                    data.insert(
                        "direction".to_string(),
                        Value::String(direction.to_string()),
                    );
                    data.insert("findings".to_string(), Value::Array(finding_dicts));
                    data.insert(
                        "mode".to_string(),
                        Value::String("block".to_string()),
                    );
                    bus.publish(EventType::SecurityBlock, data);
                }
                Err(OpenJarvisError::Security(
                    openjarvis_core::error::SecurityError::Blocked(format!(
                        "Security scan blocked {}: {} finding(s) detected",
                        direction,
                        result.findings.len()
                    )),
                ))
            }
        }
    }
}

#[async_trait::async_trait]
impl<E: InferenceEngine> InferenceEngine for GuardrailsEngine<E> {
    fn engine_id(&self) -> &str {
        self.engine.engine_id()
    }

    fn generate(
        &self,
        messages: &[Message],
        model: &str,
        temperature: f64,
        max_tokens: i64,
        extra: Option<&Value>,
    ) -> Result<GenerateResult, OpenJarvisError> {
        if self.scan_input {
            for msg in messages {
                if !msg.content.is_empty() {
                    let result = self.scan_text(&msg.content);
                    if !result.clean() {
                        self.handle_findings(&msg.content, &result, "input")?;
                    }
                }
            }
        }

        let mut response = self
            .engine
            .generate(messages, model, temperature, max_tokens, extra)?;

        if self.scan_output && !response.content.is_empty() {
            let result = self.scan_text(&response.content);
            if !result.clean() {
                response.content =
                    self.handle_findings(&response.content, &result, "output")?;
            }
        }

        Ok(response)
    }

    async fn stream(
        &self,
        messages: &[Message],
        model: &str,
        temperature: f64,
        max_tokens: i64,
        extra: Option<&Value>,
    ) -> Result<TokenStream, OpenJarvisError> {
        self.engine
            .stream(messages, model, temperature, max_tokens, extra)
            .await
    }

    fn list_models(&self) -> Result<Vec<String>, OpenJarvisError> {
        self.engine.list_models()
    }

    fn health(&self) -> bool {
        self.engine.health()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    struct MockEngine;

    #[async_trait::async_trait]
    impl InferenceEngine for MockEngine {
        fn engine_id(&self) -> &str {
            "mock"
        }
        fn generate(
            &self,
            _messages: &[Message],
            model: &str,
            _temperature: f64,
            _max_tokens: i64,
            _extra: Option<&Value>,
        ) -> Result<GenerateResult, OpenJarvisError> {
            Ok(GenerateResult {
                content: "Response with sk-test1234567890abcdefghij".into(),
                model: model.into(),
                ..Default::default()
            })
        }
        async fn stream(
            &self,
            _messages: &[Message],
            _model: &str,
            _temperature: f64,
            _max_tokens: i64,
            _extra: Option<&Value>,
        ) -> Result<TokenStream, OpenJarvisError> {
            Ok(Box::pin(futures::stream::empty()))
        }
        fn list_models(&self) -> Result<Vec<String>, OpenJarvisError> {
            Ok(vec!["mock-model".into()])
        }
        fn health(&self) -> bool {
            true
        }
    }

    #[test]
    fn test_guardrails_warn_mode() {
        let engine = MockEngine;
        let guardrails =
            GuardrailsEngine::new(engine, RedactionMode::Warn, false, true, None);
        let result = guardrails
            .generate(&[Message::user("Hi")], "mock", 0.7, 100, None)
            .unwrap();
        assert!(result.content.contains("sk-test"));
    }

    #[test]
    fn test_guardrails_redact_mode() {
        let engine = MockEngine;
        let guardrails =
            GuardrailsEngine::new(engine, RedactionMode::Redact, false, true, None);
        let result = guardrails
            .generate(&[Message::user("Hi")], "mock", 0.7, 100, None)
            .unwrap();
        assert!(result.content.contains("[REDACTED:"));
        assert!(!result.content.contains("sk-test"));
    }

    #[test]
    fn test_guardrails_block_mode() {
        let engine = MockEngine;
        let guardrails =
            GuardrailsEngine::new(engine, RedactionMode::Block, false, true, None);
        let err = guardrails
            .generate(&[Message::user("Hi")], "mock", 0.7, 100, None)
            .unwrap_err();
        assert!(matches!(err, OpenJarvisError::Security(_)));
    }
}
